src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+            scope: ["https://www.googleapis.com/auth/cloud-platform".to_string()].to_vec(),
+            subject_token: "an_example_token".to_string(),
+            subject_token_type: JWT_TOKEN_TYPE.to_string(),
+            ..ExchangeTokenRequest::default()
+        };
+        let handler = STSHandler::new();
+        let resp = handler.exchange_token(token_req).await?;
+
+        assert_eq!(
+            resp,
+            TokenResponse {
+                access_token: "an_example_token".to_string(),
+                refresh_token: None,
+                issued_token_type: ACCESS_TOKEN_TYPE.to_string(),
+                token_type: "Bearer".to_string(),
+                expires_in: 3600,
+                scope: "https://www.googleapis.com/auth/cloud-platform".to_string(),
+            }
+        );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn exchange_token_err() -> TestResult {
+        let authentication = ClientAuthentication {
+            auth_style: ClientAuthStyle::InHeader,
+            client_id: None,
+            client_secret: None,
+        };
+        let response_body = r#"{"error":"bad request"}"#;
+
+        let server = Server::run();
+        server.expect(
+            Expectation::matching(all_of![
+                request::method_path("POST", "/fail"),
+                request::body(url_decoded(contains((
+                    "grant_type",
+                    TOKEN_EXCHANGE_GRANT_TYPE
+                )))),
+                request::body(url_decoded(contains(("subject_token", "an_example_token")))),
+                request::body(url_decoded(contains((
+                    "requested_token_type",
+                    ACCESS_TOKEN_TYPE
+                )))),
+                request::body(url_decoded(contains((
+                    "subject_token_type",
+                    JWT_TOKEN_TYPE
+                )))),
+                request::headers(contains((
+                    "content-type",
+                    "application/x-www-form-urlencoded"
+                ))),
+            ])
+            .respond_with(status_code(400).body(response_body)),
+        );
+
+        let url = server.url("/fail").to_string();
+        let headers = http::HeaderMap::new();
+        let token_req = ExchangeTokenRequest {
+            url,
+            headers,
+            authentication,
+            subject_token: "an_example_token".to_string(),
+            subject_token_type: JWT_TOKEN_TYPE.to_string(),
+            ..ExchangeTokenRequest::default()
+        };
+        let handler = STSHandler::new();
+        let err = assert_err!(handler.exchange_token(token_req).await);
+
+        let expected_err = crate::errors::CredentialsError::from_str(
+            false,
+            "error exchanging token, failed with status 400 Bad Request",
+        );
+        assert_eq!(err.to_string(), expected_err.to_string());
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn refresh_access_token() -> TestResult {
```

Comment:
The implementations in the other libraries are focused on the token exchange grant type, and do not support this.

When would this be used?

Reply:
It's used in Java and Go SDKs for ExternalAccountAuthorizedUserCredentials.

* Java: https://github.com/sai-sunder-s/google-auth-library-java/blob/a7a8d7a4102b0b7c1b83791947ccb662f060eca7/oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java#L351
* Go: https://github.com/googleapis/google-cloud-go/blob/main/auth/credentials/internal/externalaccountuser/externalaccountuser.go#L91

Reply:
That is using basic auth though: e.g. https://github.com/sai-sunder-s/google-auth-library-java/blob/a7a8d7a4102b0b7c1b83791947ccb662f060eca7/oauth2_http/java/com/google/auth/oauth2/ExternalAccountAuthorizedUserCredentials.java#L365